### PR TITLE
Resolve conflicts in Lambda, Printlambda, Simplif, and Tmc

### PR DESCRIPTION
**Pull request by goldfirere**
This was originally #171, but rebased to remove lambda.mli, which was done separately in #183.

@mshinwell is the author of this code, not me.

### CODE REVIEW

**Comment by goldfirere** [comment]
Reviewing now

**Review by goldfirere — COMMENTED** [review]
Reviewed. Cannot Approve because I am the author.

**Comment by goldfirere** [comment]
lambda.ml still does not compile, because it has incomplete pattern matches in `primitive_may_allocate` and `primitive_result_layout`. Someone with knowledge of the new primitives from upstream (listed below) will need to help here. @mshinwell or designee.

New primitives:

```
Prunstack
Pperform
Presume
Preperform
Patomic_exchange
Patomic_cas
Patomic_fetch_add
Pdls_get
Patomic_load _
```

**Comment by goldfirere** [comment]
Current status:
* The first two commits here are fully reviewed.
* The third commit (very simple!) is not. Perhaps @ncik-roberts can take a very quick look.
* Still need someone who knows their way around primitives to address https://github.com/ocaml-flambda/ocaml-jst/pull/185#issuecomment-1771176667

File: lambda/printlambda.ml [hunk 1]
```diff
@@ -308,36 +292,16 @@ let primitive ppf = function
...
-  | Pfield(n, ptr, mut) ->
+  | Pfield (n, ptr, sem) ->
       let instr =
-        match ptr, mut with
+        match ptr, sem with
         | Immediate, _ -> "field_int "
-        | Pointer, Mutable -> "field_mut "
-        | Pointer, Immutable -> "field_imm "
+        | Pointer, Reads_vary -> "field_mut "
+        | Pointer, Reads_agree -> "field_imm "
```
Comment (Ekdohibs) [inline lambda/printlambda.ml:300]: This needs the trailing space removed.

**Review by Ekdohibs — APPROVED** [review]

**Comment by mshinwell** [comment]
(@lukemaurer approved the `Compilenv` parts separately)

**Pull request merged**